tsdb/db.go

```diff
@@ -760,52 +770,82 @@ func (db *DB) Compact() (err error) {
 		// consistently, we explicitly remove the last value
 		// from the block interval here.
 		head := NewRangeHead(db.head, mint, maxt-1)
-		if err := db.compactHead(head); err != nil {
+		t, err := db.compactHead(head)
+		if err != nil {
 			return err
 		}
+		if t > checkpointTime {
+			checkpointTime = t
+		}
+	}
+
+	checkpointAttempted = true
+	if err := db.performHeadCheckpoint(checkpointTime); err != nil {
+		return errors.Wrap(err, "head checkpoint in Compact")
 	}
 
 	return db.compactBlocks()
 }
 
+func (db *DB) performHeadCheckpoint(t int64) error {
+	if blocks := db.Blocks(); len(blocks) > 0 {
+		blockMaxt := blocks[len(blocks)-1].Meta().MaxTime
+		if blockMaxt > t {
+			t = blockMaxt
+		}
+	}
+
+	if t == math.MinInt64 {
```

Comment:
Uhm, zero value of `t` is `0`, not `math.MinInt64`. I think you have to initialise `t = math.MinInt64` at the beginning of the function, no?

Reply:
t is a parameter

Reply:
`t` is an argument (hence should not initialize in the method), and we are using `math.MinInt64` as zero value (which comes from `compactHead` and `Compact`)

Reply:
🤦

Reply:
As Golang's `time.Unix(x,y)` method accepts 0 and even negative numbers, I thought of keeping the zero value as minint64.